File: plugins/kinetica/skills/kinetica-execute/scripts/modules/monitor.js
'use strict';

/**
 * Monitor module for the Kinetica CLI (Node.js).
 *
 * Provides 6 commands: create, show, clear, create-trigger, clear-trigger,
 * show-triggers.
 *
 * Each export follows the category module contract:
 *   { fn: async (db, args) => void, desc: string }
 */

const { die, out, parseCsvArg, parseFloatCsv } = require('./helpers');

// ---------------------------------------------------------------------------
// Valid event types for table monitors
// ---------------------------------------------------------------------------

const EVENT_TYPES = new Set(['insert', 'update', 'delete']);

// ---------------------------------------------------------------------------
// create  --  create_table_monitor
// ---------------------------------------------------------------------------

async function cmdCreate(db, args) {
  const tableName = args.positional[0];
  if (!tableName) {
    die(
      'Usage: monitor create <table> [--event insert|update|delete] ' +
        '[--monitor-id ID] [--expression EXPR]'
    );
  }

  const event = args.flags['event'];
  if (event && !EVENT_TYPES.has(event)) {
    die(`Invalid event type '${event}'. Must be one of: insert, update, delete`);
  }

  const options = {};
  if (event) options.event = event;
  if (args.flags['monitor-id']) options.monitor_id = args.flags['monitor-id'];
  if (args.flags['datasink-name']) options.datasink_name = args.flags['datasink-name'];
  if (args.flags['expression']) options.expression = args.flags['expression'];

  const resp = await db.create_table_monitor(tableName, options);

  out({
    table_name: tableName,
    status: 'ok',
    topic_id: resp.topic_id || '',
    monitor_id: resp.monitor_id || '',
    type_schema: resp.type_schema || '',
  });
}

// ---------------------------------------------------------------------------
// show  --  show_table_monitors
// ---------------------------------------------------------------------------

async function cmdShow(db, args) {
  const monitorIds = args.flags['monitor-ids']
    ? parseCsvArg(args.flags['monitor-ids'])
    : ['*'];

  const resp = await db.show_table_monitors(monitorIds, {});

  const tableNames = resp.table_names || [];
  const resultMonitorIds = resp.monitor_ids || [];
  const types = resp.types || [];
  const events = resp.events || [];

  const monitors = resultMonitorIds.map((mid, i) => ({
    monitor_id: mid,
    table_name: i < tableNames.length ? tableNames[i] : null,
    type: i < types.length ? types[i] : null,
    event: i < events.length ? events[i] : null,
  }));

  out({ monitors, total: monitors.length });
}

// ---------------------------------------------------------------------------
// clear  --  clear_table_monitor
// ---------------------------------------------------------------------------

async function cmdClear(db, args) {
  const topicId = args.positional[0];
  if (!topicId) die('Usage: monitor clear <topic-id>');

  await db.clear_table_monitor(topicId, {});

  out({
    topic_id: topicId,
    status: 'ok',
    message: `Monitor '${topicId}' cleared`,
  });
}

// ---------------------------------------------------------------------------
// create-trigger  --  create_trigger_by_area / create_trigger_by_range
// ---------------------------------------------------------------------------

const TRIGGER_TYPES = new Set(['area', 'range']);

async function cmdCreateTrigger(db, args) {
  const tableNamesRaw = args.positional[0];
  if (!tableNamesRaw) {
    die('Usage: monitor create-trigger <table,...> --type area|range --trigger-id ID [options]');
  }

  const tableNames = parseCsvArg(tableNamesRaw);
  const triggerType = args.flags['type'];
  const triggerId = args.flags['trigger-id'];

  if (!triggerType || !TRIGGER_TYPES.has(triggerType)) {
    die('--type is required and must be one of: area, range');
  }
  if (!triggerId) {
    die('--trigger-id is required');
  }

  if (triggerType === 'area') {
    const xCol = args.flags['x-col'];
    const yCol = args.flags['y-col'];
    if (!xCol || !yCol) die('Area trigger requires --x-col and --y-col');

    const xVerticesRaw = args.flags['x-vertices'];
    const yVerticesRaw = args.flags['y-vertices'];
    if (!xVerticesRaw || !yVerticesRaw) {
      die('Area trigger requires --x-vertices and --y-vertices');
    }

    const xVector = parseFloatCsv(xVerticesRaw);
    const yVector = parseFloatCsv(yVerticesRaw);

    if (xVector.length !== yVector.length) {
      die('--x-vertices and --y-vertices must have the same number of values');
    }

    await db.create_trigger_by_area(
      triggerId,
      tableNames,
      xCol,
      xVector,
      yCol,
      yVector,
      {}
    );

    out({
      trigger_id: triggerId,
      trigger_type: 'area',
      status: 'ok',
      table_names: tableNames,
      x_column: xCol,
      y_column: yCol,
      x_vertices: xVector,
      y_vertices: yVector,
    });
  } else if (triggerType === 'range') {
    const column = args.flags['column'];
    if (!column) die('Range trigger requires --column');

    const minVal = args.flags['min'];
    const maxVal = args.flags['max'];
    if (minVal === undefined || maxVal === undefined) {
      die('Range trigger requires --min and --max');
    }

    const minNum = parseFloat(minVal);
    const maxNum = parseFloat(maxVal);

    if (Number.isNaN(minNum) || Number.isNaN(maxNum)) {
      die('--min and --max must be valid numbers');
    }

    await db.create_trigger_by_range(
      triggerId,
      tableNames,
      column,
      minNum,
      maxNum,
      {}
    );

    out({
      trigger_id: triggerId,
      trigger_type: 'range',
      status: 'ok',
      table_names: tableNames,
      column,
      min: minNum,
      max: maxNum,
    });
  }
}

// ---------------------------------------------------------------------------
// clear-trigger  --  clear_trigger
// ---------------------------------------------------------------------------

async function cmdClearTrigger(db, args) {
  const triggerId = args.positional[0];
  if (!triggerId) die('Usage: monitor clear-trigger <trigger-id>');

  await db.clear_trigger(triggerId, {});

  out({
    trigger_id: triggerId,
    status: 'ok',
    message: `Trigger '${triggerId}' cleared`,
  });
}

// ---------------------------------------------------------------------------
// show-triggers  --  show_triggers
// ---------------------------------------------------------------------------

async function cmdShowTriggers(db, args) {
  const triggerIds = args.flags['trigger-ids']
    ? parseCsvArg(args.flags['trigger-ids'])
    : ['*'];

  const resp = await db.show_triggers(triggerIds, {});

  const triggerMap = resp.trigger_map || {};

  const triggers = Object.entries(triggerMap).map(([tid, info]) => ({
    trigger_id: tid,
    info,
  }));

  out({ triggers, total: triggers.length });
}

// ---------------------------------------------------------------------------
// Exports
// ---------------------------------------------------------------------------

module.exports = {
  create: { fn: cmdCreate, desc: 'Create a table monitor' },
  show: { fn: cmdShow, desc: 'Show table monitors' },
  clear: { fn: cmdClear, desc: 'Clear (remove) a table monitor' },
  'create-trigger': {
    fn: cmdCreateTrigger,
    desc: 'Create a trigger (area or range) on tables',
  },
  'clear-trigger': {
    fn: cmdClearTrigger,
    desc: 'Clear (remove) a trigger',
  },
  'show-triggers': {
    fn: cmdShowTriggers,
    desc: 'Show triggers',
  },
};

File: skills/kinetica-execute/scripts/modules/monitor.py
"""Monitor module for the Kinetica CLI (Python).

Provides 6 commands: create, show, clear, create-trigger, clear-trigger,
show-triggers.

Each command follows the category module contract:
    COMMANDS = { "name": {"fn": callable, "desc": str, "build_args": callable} }
"""

from modules.helpers import check_status, die, out, parse_csv_arg, parse_float_csv


# ---------------------------------------------------------------------------
# Valid event types for table monitors
# ---------------------------------------------------------------------------

_EVENT_TYPES = frozenset(["insert", "update", "delete"])


# ---------------------------------------------------------------------------
# create  --  create_table_monitor
# ---------------------------------------------------------------------------

def _build_create_args(parser):
    parser.add_argument("table_name", help="Table to monitor")
    parser.add_argument(
        "--event", default=None, choices=sorted(_EVENT_TYPES),
        help="Event type to monitor (insert, update, or delete)",
    )
    parser.add_argument("--monitor-id", dest="monitor_id", default=None, help="Custom monitor ID")
    parser.add_argument("--datasink-name", dest="datasink_name", default=None, help="Datasink name to route events to")
    parser.add_argument("--expression", default=None, help="Filter expression for monitored events")


def cmd_create(db, args):
    table_name = args.table_name
    if not table_name:
        die("Usage: monitor create <table> [--event insert|update|delete] "
            "[--monitor-id ID] [--expression EXPR]")

    options = {}
    if args.event:
        options["event"] = args.event
    if args.monitor_id:
        options["monitor_id"] = args.monitor_id
    if args.datasink_name:
        options["datasink_name"] = args.datasink_name
    if args.expression:
        options["expression"] = args.expression

    resp = db.create_table_monitor(
        table_name=table_name,
        options=options,
    )
    check_status(resp, "create_table_monitor")

    out({
        "table_name": table_name,
        "status": "ok",
        "topic_id": resp.get("topic_id", ""),
        "monitor_id": resp.get("monitor_id", ""),
        "type_schema": resp.get("type_schema", ""),
    })


# ---------------------------------------------------------------------------
# show  --  show_table_monitors
# ---------------------------------------------------------------------------

def _build_show_args(parser):
    parser.add_argument(
        "--monitor-ids", dest="monitor_ids", default=None,
        help="Comma-separated monitor IDs to show (omit to list all)",
    )


def cmd_show(db, args):
    monitor_ids = parse_csv_arg(args.monitor_ids) if args.monitor_ids else ["*"]

    resp = db.show_table_monitors(
        monitor_ids=monitor_ids,
        options={},
    )
    check_status(resp, "show_table_monitors")

    # Build a structured list from the response
    table_names = resp.get("table_names", [])
    result_monitor_ids = resp.get("monitor_ids", [])
    types = resp.get("types", [])
    events = resp.get("events", [])

    monitors = []
    for i, mid in enumerate(result_monitor_ids):
        monitors.append({
            "monitor_id": mid,
            "table_name": table_names[i] if i < len(table_names) else None,
            "type": types[i] if i < len(types) else None,
            "event": events[i] if i < len(events) else None,
        })

    out({"monitors": monitors, "total": len(monitors)})


# ---------------------------------------------------------------------------
# clear  --  clear_table_monitor
# ---------------------------------------------------------------------------

def _build_clear_args(parser):
    parser.add_argument("topic_id", help="Topic ID of the monitor to clear")


def cmd_clear(db, args):
    topic_id = args.topic_id
    if not topic_id:
        die("Usage: monitor clear <topic-id>")

    resp = db.clear_table_monitor(
        topic_id=topic_id,
        options={},
    )
    check_status(resp, "clear_table_monitor")

    out({
        "topic_id": topic_id,
        "status": "ok",
        "message": f"Monitor '{topic_id}' cleared",
    })


# ---------------------------------------------------------------------------
# create-trigger  --  create_trigger_by_area / create_trigger_by_range
# ---------------------------------------------------------------------------

_TRIGGER_TYPES = frozenset(["area", "range"])


def _build_create_trigger_args(parser):
    parser.add_argument("table_names", help="Comma-separated table names to attach the trigger to")
    parser.add_argument("--type", dest="trigger_type", required=True, choices=sorted(_TRIGGER_TYPES), help="Trigger type (area or range)")
    parser.add_argument("--trigger-id", dest="trigger_id", required=True, help="User-provided trigger ID")

    # Area-specific arguments
    parser.add_argument("--x-col", dest="x_col", default=None, help="X column name (area trigger)")
    parser.add_argument("--y-col", dest="y_col", default=None, help="Y column name (area trigger)")
    parser.add_argument("--x-vertices", dest="x_vertices", default=None, help="Comma-separated X polygon vertices (area trigger)")
    parser.add_argument("--y-vertices", dest="y_vertices", default=None, help="Comma-separated Y polygon vertices (area trigger)")

    # Range-specific arguments
    parser.add_argument("--column", default=None, help="Column name (range trigger)")
    parser.add_argument("--min", dest="range_min", type=float, default=None, help="Minimum value (range trigger)")
    parser.add_argument("--max", dest="range_max", type=float, default=None, help="Maximum value (range trigger)")


def cmd_create_trigger(db, args):
    table_names = parse_csv_arg(args.table_names)
    trigger_type = args.trigger_type
    trigger_id = args.trigger_id

    if not table_names:
        die("Usage: monitor create-trigger <table,...> --type area|range --trigger-id ID [options]")

    if trigger_type == "area":
        if not args.x_col or not args.y_col:
            die("Area trigger requires --x-col and --y-col")
        if not args.x_vertices or not args.y_vertices:
            die("Area trigger requires --x-vertices and --y-vertices")

        x_vector = parse_float_csv(args.x_vertices)
        y_vector = parse_float_csv(args.y_vertices)

        if len(x_vector) != len(y_vector):
            die("--x-vertices and --y-vertices must have the same number of values")

        resp = db.create_trigger_by_area(
            request_id=trigger_id,
            table_names=table_names,
            x_column_name=args.x_col,
            x_vector=x_vector,
            y_column_name=args.y_col,
            y_vector=y_vector,
            options={},
        )
        check_status(resp, "create_trigger_by_area")

        out({
            "trigger_id": trigger_id,
            "trigger_type": "area",
            "status": "ok",
            "table_names": table_names,
            "x_column": args.x_col,
            "y_column": args.y_col,
            "x_vertices": x_vector,
            "y_vertices": y_vector,
        })

    elif trigger_type == "range":
        if not args.column:
            die("Range trigger requires --column")
        if args.range_min is None or args.range_max is None:
            die("Range trigger requires --min and --max")

        resp = db.create_trigger_by_range(
            request_id=trigger_id,
            table_names=table_names,
            column_name=args.column,
            min=args.range_min,
            max=args.range_max,
            options={},
        )
        check_status(resp, "create_trigger_by_range")

        out({
            "trigger_id": trigger_id,
            "trigger_type": "range",
            "status": "ok",
            "table_names": table_names,
            "column": args.column,
            "min": args.range_min,
            "max": args.range_max,
        })


# ---------------------------------------------------------------------------
# clear-trigger  --  clear_trigger
# ---------------------------------------------------------------------------

def _build_clear_trigger_args(parser):
    parser.add_argument("trigger_id", help="Trigger ID to clear")


def cmd_clear_trigger(db, args):
    trigger_id = args.trigger_id
    if not trigger_id:
        die("Usage: monitor clear-trigger <trigger-id>")

    resp = db.clear_trigger(
        trigger_id=trigger_id,
        options={},
    )
    check_status(resp, "clear_trigger")

    out({
        "trigger_id": trigger_id,
        "status": "ok",
        "message": f"Trigger '{trigger_id}' cleared",
    })


# ---------------------------------------------------------------------------
# show-triggers  --  show_triggers
# ---------------------------------------------------------------------------

def _build_show_triggers_args(parser):
    parser.add_argument(
        "--trigger-ids", dest="trigger_ids", default=None,
        help="Comma-separated trigger IDs to show (omit to list all)",
    )


def cmd_show_triggers(db, args):
    trigger_ids = parse_csv_arg(args.trigger_ids) if args.trigger_ids else ["*"]

    resp = db.show_triggers(
        trigger_ids=trigger_ids,
        options={},
    )
    check_status(resp, "show_triggers")

    trigger_map = resp.get("trigger_map", {})

    triggers = []
    for tid, info in trigger_map.items():
        triggers.append({
            "trigger_id": tid,
            "info": info,
        })

    out({"triggers": triggers, "total": len(triggers)})


# ---------------------------------------------------------------------------
# COMMANDS dict
# ---------------------------------------------------------------------------

COMMANDS = {
    "create": {
        "fn": cmd_create,
        "desc": "Create a table monitor",
        "build_args": _build_create_args,
    },
    "show": {
        "fn": cmd_show,
        "desc": "Show table monitors",
        "build_args": _build_show_args,
    },
    "clear": {
        "fn": cmd_clear,
        "desc": "Clear (remove) a table monitor",
        "build_args": _build_clear_args,
    },
    "create-trigger": {
        "fn": cmd_create_trigger,
        "desc": "Create a trigger (area or range) on tables",
        "build_args": _build_create_trigger_args,
    },
    "clear-trigger": {
        "fn": cmd_clear_trigger,
        "desc": "Clear (remove) a trigger",
        "build_args": _build_clear_trigger_args,
    },
    "show-triggers": {
        "fn": cmd_show_triggers,
        "desc": "Show triggers",
        "build_args": _build_show_triggers_args,
    },
}
